File: server/src/main/java/io/fnx/backend/domain/FileEntity.java
package io.fnx.backend.domain;

import com.google.common.collect.Sets;
import com.googlecode.objectify.Key;
import com.googlecode.objectify.annotation.Entity;
import com.googlecode.objectify.annotation.Id;
import com.googlecode.objectify.annotation.Index;
import com.googlecode.objectify.annotation.Unindex;
import org.joda.time.DateTime;

import java.util.Collections;
import java.util.Set;

import static io.fnx.backend.tools.ofy.OfyUtils.nameToKey;
import static java.lang.String.format;

@Entity
@Unindex
public class FileEntity {

    public static final Set<String> IMAGE_MEDIA_TYPES = Collections.unmodifiableSet(
            Sets.newHashSet("image/jpeg", "image/png")
    );

    @Id
    private String id;

    private String name;
    private String bucketUrl;
    private String imageUrl;

    @Index
    private FileCategory category;

    @Index
    private String set; //a.k.a. folder!

    private Key<UserEntity> uploader;
    @Index
    private DateTime uploaded;
    private String mediaType;

    public static Key<FileEntity> createKey(String id) {
        return nameToKey(FileEntity.class, id);
    }

    public Key<FileEntity> getKey() {
        return createKey(id);
    }

    public void setId(String filePrefix, String fileName) {
        this.id = format("%s/%s", filePrefix, fileName);
    }

    public boolean isImage() {
        return IMAGE_MEDIA_TYPES.contains(mediaType);
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getBucketUrl() {
        return bucketUrl;
    }

    public void setBucketUrl(String bucketUrl) {
        this.bucketUrl = bucketUrl;
    }

    public String getImageUrl() {
        return imageUrl;
    }

    public void setImageUrl(String imageUrl) {
        this.imageUrl = imageUrl;
    }

    public FileCategory getCategory() {
        return category;
    }

    public void setCategory(FileCategory category) {
        this.category = category;
    }

    public Key<UserEntity> getUploader() {
        return uploader;
    }

    public void setUploader(Key<UserEntity> uploader) {
        this.uploader = uploader;
    }

    public DateTime getUploaded() {
        return uploaded;
    }

    public void setUploaded(DateTime uploaded) {
        this.uploaded = uploaded;
    }

    public void setMediaType(String mediaType) {
        this.mediaType = mediaType;
    }

    public String getMediaType() {
        return mediaType;
    }

	public String getSet() {
		return set;
	}

	public void setSet(String set) {
		this.set = set;
	}
}

File: api_client/lib/model/file_entity.dart
part of api_client.api;

class FileEntity {
  int id = null;

  String name = null;

  String bucketUrl = null;

  String imageUrl = null;

  FileCategory category = null;

/* folder */
  String set_ = null;

  Key uploader = null;

  DateTime uploaded = null;

  String mediaType = null;

  FileEntity();

  @override
  String toString() {
    return 'FileEntity[id=$id, name=$name, bucketUrl=$bucketUrl, imageUrl=$imageUrl, category=$category, set_=$set_, uploader=$uploader, uploaded=$uploaded, mediaType=$mediaType, ]';
  }

  FileEntity.fromJson(Map<String, dynamic> json) {
    if (json == null) return;
    id = json['id'];
    name = json['name'];
    bucketUrl = json['bucketUrl'];
    imageUrl = json['imageUrl'];
    category = new FileCategory.fromJson(json['category']);
    set_ = json['set_'];
    uploader = new Key.fromJson(json['uploader']);
    uploaded =
        json['uploaded'] == null ? null : DateTime.parse(json['uploaded']);
    mediaType = json['mediaType'];
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'bucketUrl': bucketUrl,
      'imageUrl': imageUrl,
      'category': category.toJson(),
      'set_': set_,
      'uploader': uploader.toJson(),
      'uploaded': uploaded == null ? '' : uploaded.toUtc().toIso8601String(),
      'mediaType': mediaType
    };
  }

  static List<FileEntity> listFromJson(List<dynamic> json) {
    return json == null
        ? new List<FileEntity>()
        : json.map((value) => new FileEntity.fromJson(value)).toList();
  }

  static Map<String, FileEntity> mapFromJson(
      Map<String, Map<String, dynamic>> json) {
    var map = new Map<String, FileEntity>();
    if (json != null && json.length > 0) {
      json.forEach((String key, Map<String, dynamic> value) =>
          map[key] = new FileEntity.fromJson(value));
    }
    return map;
  }
}
